library/alloc/src/vec/in_place_collect.rs

```diff
@@ -193,12 +196,16 @@ where
 
         // Drop any remaining values at the tail of the source but prevent drop of the allocation
         // itself once IntoIter goes out of scope.
-        // If the drop panics then we also leak any elements collected into dst_buf.
+        // If the drop panics then we also try to drop the destination buffer and its elements.
+        // This is safe because `forget_allocation_drop_remaining` immediately forgets the allocation
+        // and won't panic before that.
         //
         // Note: This access to the source wouldn't be allowed by the TrustedRandomIteratorNoCoerce
         // contract (used by SpecInPlaceCollect below). But see the "O(1) collect" section in the
         // module documenttation why this is ok anyway.
+        let dst_guard = InPlaceDstBufDrop { ptr: dst_buf, len, cap };
```

Comment:
I'm afraid this will lead to type confusion.`dst_buf` is `T` aka `New` but we ought to drop them as `Old`.

Reply:
`src.forget_allocation_drop_remaining()` is the one that drops the old elements.
`dst_guard` is meant to drop the new elements (and the allocation) in case dropping an old element panics.

I see however the potential for confusion since the original docs were meant to annotate just the `src.forget_allocation_drop_remaining()`